src/middlewared/middlewared/plugins/jbof/redfish/client.py

```diff
@@ -341,3 +347,258 @@ def link_status(self, uri):
+        # Clear any silenced exception logging
+        try:
+            self.logged.remove(base_url)
+        except KeyError:
+            pass
+        return newsession
+
+    async def _logout(self, base_url, session):
+        if self.authtype == AuthMethod.BASIC:
+            self.auth = None
+        elif self.authtype == AuthMethod.SESSION:
+            if location := self.session_location.get(base_url):
+                async with session.delete(location, ssl=False) as response:
+                    if response.ok:
+                        try:
+                            del self.session_location[base_url]
+                        except KeyError:
+                            # Should not occur, but protect in case parallel calls
+                            pass
+
+    async def get_root_object(self):
+        """Fetch the root object."""
+        # We're not going to try to be clever about which base_urls are used for this.
+        base_url_count = len(self.base_urls)
+        for index, base_url in enumerate(self.base_urls, 1):
+            try:
+                async with aiohttp.ClientSession(base_url, timeout=self.timeout, raise_for_status=True) as session:
+                    async with session.get(self.prefix, ssl=False) as response:
+                        return await response.json()
+            except asyncio.TimeoutError:
+                if index == base_url_count:
+                    raise CallError('Connection timed out', errno.ETIMEDOUT)
+                else:
+                    continue
+            except Exception:
+                continue
+
+        raise CallError('Failed to obtain root object', errno.EBADMSG)
+
+    async def get(self, uri):
+        if not uri.startswith(self.prefix):
+            uri = f'{self.prefix}{uri}'
+        # Iterate over the available paths
+        for base_url, session in self.sessions():
+            try:
+                if not session:
+                    session = await self._login(base_url)
+                async with session.get(uri, ssl=False) as response:
+                    if response.ok:
+                        return await response.json()
+            except asyncio.TimeoutError:
+                LOGGER.debug('Timed out GET %r: %r', base_url, uri)
+                await self._del_session(base_url, session)
+                continue
+            except Exception:
+                if base_url not in self.logged:
+                    LOGGER.debug('Failed GET %r: %r', base_url, uri, exc_info=True)
+                    self.logged.add(base_url)
+                await self._del_session(base_url, session)
+                continue
+        raise CallError(f'Failed to GET {uri}:', errno.EBADMSG)
+
+    async def post(self, uri, **kwargs):
+        if not uri.startswith(self.prefix):
+            uri = f'{self.prefix}{uri}'
+        payload = kwargs.get('data', {})
+        # Iterate over the available paths
+        for base_url, session in self.sessions():
```

Comment:
Maybe it's because there isn't any consumer of this new class but I'm failing to understand how the `post` and `get` methods work. If I'm reading it correctly, the methods are given a URI and then we iterate over all cached sessions and on the first session for which we get a valid response, we short-circuit and return the information. So this means we return the information on the 1st successful cached session?

Reply:
Correct.  The idea is that each `base_url` represents a path by which an operation may be accomplished (GET, POST, etc) - we don't care which path gets used, but we want to be resilient in case one path goes down.

In the `cache_get` classmethod you can see how `base_urls` gets set to `['https://<mgmt_ip1>', 'https://<mgmt_ip2>']` and (in `_login`) we construct `aiohttp.ClientSession` corresponding to each (well, we're lazy so we may never actually construct the 2nd if the 1st works all the time).

FWIW, the `chassis` and `managers` methods indirectly call the `get`.